pkg/cloud/azure/azure.go

```diff
@@ -85,3 +110,71 @@ func NewProviderAssets(config config.OperatorConfig) (common.CloudProviderAssets
 	}
 	return assets, nil
 }
+
+func IsAzure(infra *configv1.Infrastructure) bool {
+	if infra.Status.PlatformStatus != nil {
+		if infra.Status.PlatformStatus.Type == configv1.AzurePlatformType &&
+			(infra.Status.PlatformStatus.Azure.CloudName != configv1.AzureStackCloud) {
+			return true
+		}
+	}
+	return false
+}
+
+func CloudConfigTransformer(source string, infra *configv1.Infrastructure, network *configv1.Network) (string, error) {
+	if !IsAzure(infra) {
+		return "", fmt.Errorf("invalid platform, expected CloudName to be %s", configv1.AzurePublicCloud)
+	}
+
+	var cfg azure.Config
+	if err := json.Unmarshal([]byte(source), &cfg); err != nil {
+		return "", fmt.Errorf("failed to unmarshal the cloud.conf: %w", err)
+	}
+
+	// We are copying the behaviour from CCO's transformer we need to:
+	// 1. Ensure that the Cloud is set in the cloud.conf
+	//   i. If it is set, verify that it is valid and does not conflict with the
+	//      infrastructure config. If it conflicts, we want to error
+	//  ii. If it is not set, default to public cloud (configv1.AzurePublicCloud)
+	//
+	// 2. Verify the cloud name set in the infra config is valid, if it is not
+	// bail with an informative error
+
+	// Verify the cloud name set in the infra config is valid
+	cloud := configv1.AzurePublicCloud
+	if azurePlatform := infra.Status.PlatformStatus.Azure; azurePlatform != nil {
+		if c := azurePlatform.CloudName; c != "" {
+			if !validAzureCloudNames[c] {
+				return "", field.NotSupported(field.NewPath("status", "platformStatus", "azure", "cloudName"), c, validAzureCloudNameValues)
+			}
+			cloud = c
+		}
+	}
+
+	// Ensure cloud set in cloud.conf matches infra
+	if cfg.Cloud != "" {
+		if !strings.EqualFold(string(cloud), cfg.Cloud) {
+			return "",
+				fmt.Errorf(`invalid user-provided cloud.conf: \"cloud\" field in user-provided
+				cloud.conf conflicts with infrastructure object`)
+		}
+	}
+
+	cfg.Cloud = string(cloud)
+
+	// TODO: Remove when you work this out (before merging)
+	// Why is cfg.Cloud not typed: type AzureCloudEnvironment string
```

Comment:
Is the cfg.Cloud an upstream struct definition or something we have locally? If it's an upstream definition, I guess there isn't much we can do

Reply:
Ah it's upstream. Cheers! :D